vermouth/map_input.py

```diff
@@ -20,28 +20,159 @@
+                mappings[from_ff][to_ff][name] = (mapping, weights, extra)
+        else:
+            break
+
+    # We do not want to return defaultdicts.
+    mappings = _default_to_dict(mappings)
+
+    return mappings
+
+
+def _default_to_dict(mappings):
+    """
+    Convert a mapping collection from a defaultdict to a dict.
+    """
+    new_mappings = {}
+    for from_ff, from_dict in mappings.items():
+        new_mappings[from_ff] = {}
+        for to_ff, to_dict in from_dict.items():
+            new_mappings[from_ff][to_ff] = dict(to_dict)
+    return new_mappings
+
+
+def _compute_weights(mapping, name):
+    """
+    Calculate the mapping weights from a preliminary mapping.
+
+    The preliminary mapping refers to atoms by their names instead of the tuple
+    ``(resid, name)`` used in a final mapping. Target atoms with null weights
+    are prefixed with a '!'. The result dictionary also refers to atoms
+    directly by name.
+
+    Parameters
+    ----------
+    mapping: dict[str, list[str]]
+        Preliminary mapping.
+    name: str
+        Molecule name. Used to generate error messages.
+
+    Returns
+    -------
+    dict
+    """
+    rev_mapping = collections.defaultdict(list)
+    for from_atom, to_atoms in mapping.items():
+        for to_atom in to_atoms:
+            rev_mapping[to_atom].append(from_atom)
+
+    # Atoms can be mapped with a null weight by prefixing the target particle
+    # with a "!". We first set the non-null weights.
+    pre_weights = {
+        from_atom: dict(collections.Counter([
+            to_atom for to_atom in to_atoms if not to_atom.startswith('!')
+        ]))
+        for from_atom, to_atoms in mapping.items()
+    }
+    for atom_weights in pre_weights.values():
+        total = sum(atom_weights.values())
+        for to_atom in atom_weights:
+            atom_weights[to_atom] /= total
+    weights = collections.defaultdict(dict)
+    for from_atom, to_atoms in pre_weights.items():
+        for to_atom, weight in to_atoms.items():
+            weights[to_atom][from_atom] = weight
+
+    # Then we add the null weights.
+    null_weights = {
+        to_atom[1:]: {from_atom: 0 for from_atom in from_atoms}
+        for to_atom, from_atoms in rev_mapping.items()
+        if to_atom.startswith('!')
+    }
+    for to_atom, from_weights in null_weights.items():
+        null_keys = set(from_weights.keys())
+        non_null_keys = set(weights.get(to_atom, {}).keys())
+        redifined_keys = null_keys & non_null_keys
+        if redifined_keys:
+            msg = ('Atom(s) {} is mapped to "{}" with and without a weight '
+                   'in the molecule "{}". '
+                   'There cannot be the same target atom name with and '
+                   'without a "!" prefix on a same line.')
+            raise IOError(msg.format(redifined_keys, to_atom, name))
+        weights[to_atom] = weights.get(to_atom, {})
```

Comment:
You do this get twice. Can you combine it?

Reply:
Maybe, but it would need to scramble the current logic which is very easy to get wrong.